### ISSUE TITLE
Why do I keep getting stuck here when using AdversarialTrainer ? Where is the problem?

### ISSUE THREAD

**Issue by LxyoryxL** [Trusted-AI/adversarial-robustness-toolbox#1057]
# **here is my error:**
_Precompute adv samples:   0%|          | 0/1 [00:00<?, ?it/s]
Adversarial training epochs:   0%|          | 0/20 [00:00<?, ?it/s]
/home/data/xinyuan/.conda/envs/adv_rob_lxy/lib/python3.7/site-packages/art/estimators/classification/pytorch.py:694: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  x_grad = torch.tensor(x).to(self._device)
/home/data/xinyuan/.conda/envs/adv_rob_lxy/lib/python3.7/site-packages/art/estimators/classification/pytorch.py:695: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  y_grad = torch.tensor(y).to(self._device)_

# here is my code:

### Step 0: Define the neural network model, return logits instead of activation in forward method
_class CNN_model1(nn.Module):
    def __init__(self):
        super(CNN_model1, self).__init__()
        self.conv1=nn.Sequential(
            nn.Conv2d(kernel_size=(3,3), in_channels=3, out_channels=32),
            nn.BatchNorm2d(32), 
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2)
        )
        self.conv2=nn.Sequential(
            nn.Conv2d(kernel_size=(3,3), in_channels=32, out_channels=64),
            nn.BatchNorm2d(64), 
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2)
        )
        self.fc1=nn.Sequential(
            nn.Linear(54*54*64, 64),
            nn.ReLU()
        )
        self.fc2=nn.Sequential(
            nn.Linear(64, 2),
        )
    def forward(self, x):
        x=self.conv1(x)
        x=self.conv2(x)
        x=x.contiguous().view(x.size(0), -1)# 扁平化
        x=self.fc1(x)
        x=self.fc2(x)
        return x_

### Step 1: Load the dataset
_train_datagen=ImageDataGenerator(rescale=1./255)
train_loader=train_datagen.flow_from_directory("./dataset/ALL_IDB2_origin" ,
                                                        target_size=(224, 224),
                                                        batch_size=batch_size,
                                                        shuffle=True)

x_train=np.zeros(shape=(nTrain, 224, 224, 3))
y_train=np.zeros(shape=(nTrain, 2))

i=0
for inputs,target in train_loader:
    x_train[i * batch_size : (i + 1) * batch_size] = inputs
    y_train[i * batch_size : (i + 1) * batch_size] = target
    i=i+1
    if i*batch_size>=nTrain:
        break_

### Step 1a: Swap axes to PyTorch's NCHW format

_x_train = np.transpose(x_train, (0, 3, 1, 2)).astype(np.float32)
print(x_train.shape)_

### Step 2: Create the model

_cnn_model1=CNN_model1()_

### Step 2a: Define the loss function and the optimizer

_criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(cnn_model1.parameters(), lr=0.001)_

### Step 3: Create the ART classifier

_classifier = PyTorchClassifier(
    model=cnn_model1,
    loss=criterion,
    optimizer=optimizer,
    input_shape=(3, 224, 224),
    nb_classes=2,
    device_type='gpu'
)_

### Step 4: Train the ART classifier
_classifier.fit(x_train, y_train, batch_size=10, nb_epochs=10)_

### Step 5: Adversarial Train
_from art.defences.trainer import AdversarialTrainer
from art.attacks.evasion import BasicIterativeMethod

attacks = BasicIterativeMethod(classifier, eps=0.3, eps_step=0.01, max_iter=40)
trainer = AdversarialTrainer(classifier, attacks, ratio=1.0)
trainer.fit(x_train, y_train, nb_epochs=20, batch_size=1)_

**Comment by beat-buesser** [comment]
Hi @LxyoryxL 

Thank you very much for using ART! Which version of ART and `torch` are you using?

**Comment by beat-buesser** [comment]
@LxyoryxL I have been able to reproduce you example. The warning message at the top is a `UserWarning` which we'll address in our next patch release. The adversarial trainer is slow because of the small `batch_size=1`, I would recommend to use a larger batch size to make better use of your GPU. I also would recommend to use `ProjectedGradientDescent` instead of `BasicIterativeMethod` because ART has a PGD version optimised for PyTorch.
I don't think you need Step 4 if you plan to reproduce adversarial training, just start the `AdversarialTrainer` with a randomly initialised model in `classifier`.

**Comment by LxyoryxL** [comment]
> @LxyoryxL I have been able to reproduce you example. The warning message at the top is a `UserWarning` which we'll address in our next patch release 1.6.2.
> 
> The adversarial trainer is slow because of the small `batch_size=1`, I would recommend to use a larger batch size to make better use of your GPU. I also would recommend to use `ProjectedGradientDescent` instead of `BasicIterativeMethod` because ART has a PGD version optimised for PyTorch which should run much faster.
> 
> I don't think you need Step 4 if you plan to reproduce adversarial training, just start the `AdversarialTrainer` with a randomly initialised model in `classifier`.

Why should we directly use randomly initialized models for adversarial training? Shouldn't it be an attack on an already trained model? Then attack the model to get the adversarial sample, and then add the adversarial sample to the original sample to perform one or more rounds of training again. This process is equivalent to fine-tuning the weight of the model

**Comment by beat-buesser** [comment]
Hi @LxyoryxL Please apologise, I have missed your message. `AdversarialTrainer` will train your model, then you can attack that train model which will be more robust because of the adversarial training. Starting adversarial training from random initialisation if often easier than starting from model trained first on benign samples.
